massa-network/src/peer_info_database.rs

```diff
@@ -847,47 +768,345 @@ impl PeerInfoDatabase {
-                    banned: false,
-                    bootstrap: false,
-                    last_alive: None,
-                    last_failure: None,
-                    advertised: false,
-                    active_out_connection_attempts: 0,
-                    active_out_connections: 0,
-                    active_in_connections: 0,
-                };
-                if self.active_in_nonbootstrap_connections
-                    >= self.network_settings.max_in_nonbootstrap_connections
-                {
-                    Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
-                } else if peer.active_in_connections
-                    >= self.network_settings.max_in_connections_per_ip
-                {
+                let mut peer = PeerInfo::new(*ip, false);
+                if peer.active_in_connections >= self.network_settings.max_in_connections_per_ip {
                     self.request_dump()?;
                     Err(NetworkConnectionErrorType::MaxPeersConnectionReached(*ip))
                 } else {
                     peer.active_in_connections += 1;
                     vac.insert(peer);
-                    self.active_in_nonbootstrap_connections += 1;
                     Ok(())
                 }
             }
         };
+        let peer = self.get_peer_type(ip).ok_or({
+            NetworkError::PeerConnectionError(NetworkConnectionErrorType::PeerInfoNotFoundError(
+                *ip,
+            ))
+        })?;
+
+        self.update_global_active_in_connection_count(
+            peer,
+            true,
+            NetworkConnectionErrorType::UnexpectedError,
+        )?;
         if let Err(res) = res {
             return Err(NetworkError::PeerConnectionError(res));
         }
         self.request_dump()?;
         Ok(())
     }
+
+    ////////////////////
+    // public getters //
+    ////////////////////
+
+    /// Sorts peers by ( last_failure, rev(last_success) )
+    /// and returns as many peers as there are available slots to attempt outgoing connections to.
+    pub fn get_out_connection_candidate_ips(&self) -> Result<Vec<IpAddr>, NetworkError> {
+        let mut res: Vec<_> = self
+            .get_candidate_ips_for_type(
+                PeerType::Bootstrap,
+                &self.bootstrap_connection_count,
+                &self.network_settings.bootstrap_peers_config,
+            )?
+            .chain(self.get_candidate_ips_for_type(
+                PeerType::WhiteListed,
+                &self.whitelist_connection_count,
+                &self.network_settings.whitelist_peers_config,
+            )?)
+            .chain(self.get_candidate_ips_for_type(
+                PeerType::Standard,
+                &self.standard_connection_count,
+                &self.network_settings.standard_peers_config,
+            )?)
+            .collect();
+        res.sort_unstable_by_key(|&p| (p.last_failure, std::cmp::Reverse(p.last_alive)));
```

Comment:
let's prioritize IPs of special type over standard IPs in this sorting

Reply:
I can move the sorting part in get_candidate_ips_for_type and then just chain the iterators without sorting. Is that the expected behaviour ?

Reply:
I guess just adding `p.type == standard` as the first element of the sorting tuple should do the trick

Reply:
I moved the sorting part to the for_types functions.